### PR TITLE
Add gcp_monitoring_alert_policy table. closes #145

### LINKED ISSUE #145: Add "ALERT" table under Monitoring for Metrics alerts
2.4 Ensure log metric filter and alerts exist for project ownership assignments/changes (Scored)
2.5 Ensure that the log metric filter and alerts exist for Audit Configuration changes (Scored)
2.6 Ensure that the log metric filter and alerts exist for Custom Role changes (Scored)

### PR DESCRIPTION
**Pull request by Paulami30**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT 300

SETUP: tests/gcp_monitoring_alert_policy []

PRETEST: tests/gcp_monitoring_alert_policy

TEST: tests/gcp_monitoring_alert_policy
Running terraform
data.google_client_config.current: Refreshing state...
google_monitoring_alert_policy.named_test_resource: Refreshing state... [id=projects/dummy-aaa/alertPolicies/11789617472293322293]
data.null_data_source.resource: Refreshing state...
data.template_file.resource_aka: Refreshing state...
google_monitoring_alert_policy.named_test_resource: Modifying... [id=projects/dummy-aaa/alertPolicies/11789617472293322293]
google_monitoring_alert_policy.named_test_resource: Modifications complete after 6s [id=projects/dummy-aaa/alertPolicies/11789617472293322293]
data.template_file.resource_aka: Refreshing state...

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

Outputs:

project_id = dummy-aaa
resource_aka = gcp://monitoring.googleapis.com/projects/dummy-aaa/alertPolicies/11789617472293322293
resource_id = projects/dummy-aaa/alertPolicies/11789617472293322293
resource_name = turbottest99725

Running SQL query: test-get-query.sql
[
  {
    "combiner": "OR",
    "display_name": "turbottest99725",
    "documentation": {
      "content": "This is for testing purpose",
      "mimeType": "text/markdown"
    },
    "enabled": true,
    "location": "global",
    "name": "11789617472293322293",
    "project": "dummy-aaa",
    "user_labels": {
      "foo": "bar"
    }
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "combiner": "OR",
    "display_name": "turbottest99725",
    "enabled": true,
    "name": "11789617472293322293",
    "project": "dummy-aaa",
    "tags": {
      "foo": "bar"
    }
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "combiner": "OR",
    "display_name": "turbottest99725",
    "name": "11789617472293322293"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

POSTTEST: tests/gcp_monitoring_alert_policy

TEARDOWN: tests/gcp_monitoring_alert_policy

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info
select
  display_name,
  name,
  enabled,
  documentation ->> 'content' as doc_content,
  tags
from
  gcp_monitoring_alert_policy;
  
+--------------+---------------------+---------+-------------+--------+
| display_name | name                | enabled | doc_content | tags   |
+--------------+---------------------+---------+-------------+--------+
| test-alert   | 1385080259726788053 | true    | testing     | <null> |
+--------------+---------------------+---------+-------------+--------+

### Get the creation record for each alert policy
select
  display_name,
  name,
  creation_record ->> 'mutateTime' as mutation_time,
  creation_record ->> 'mutatedBy' as mutated_by
from
  gcp_monitoring_alert_policy;
  
  +--------------+---------------------+--------------------------------+--------------------+
| display_name | name                | mutation_time                  | mutated_by         |
+--------------+---------------------+--------------------------------+--------------------+
| test-alert   | 1385080259726788053 | 2021-04-07T07:11:34.362714499Z | dummy12@turbot.com |
+--------------+---------------------+--------------------------------+--------------------+

### Get the condition details for each alert policy
select
  display_name,
  con ->> 'displayName' as filter_display_name,
  con -> 'conditionThreshold' ->> 'filter' as filter,
  con -> 'conditionThreshold' ->> 'thresholdValue' as threshold_value,
  con -> 'conditionThreshold' ->> 'trigger' as trigger
from
  gcp_monitoring_alert_policy,
  jsonb_array_elements(conditions) as con;
  
  +--------------+-----------------------------------------+----------------------------------------------------------------------------------------------------
| display_name | filter_display_name                     | filter                                                                                             
+--------------+-----------------------------------------+----------------------------------------------------------------------------------------------------
| test-alert   | VM Instance - CPU usage for test [RATE] | metric.type="compute.googleapis.com/instance/cpu/usage_time" resource.type="gce_instance" metric.la
+--------------+-----------------------------------------+----------------------------------------------------------------------------------------------------

```
</details>

### CODE REVIEW

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+			{
+				Name:        "notification_channels",
+				Description: "Identifies the notification channels to which notifications should be sent when incidents are opened or closed or when new violations occur on an already opened incident.",
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "null_fields",
+				Description: "NullFields is a list of field names (e.g. 'Combiner') to include in API requests with the JSON null value.",
+				Type:        proto.ColumnType_JSON,
+			},
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:84]: Not needed
```suggestion
```

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+			{
+				Name:        "documentation",
+				Description: "Documentation that is included with notifications and incidents related to this policy.",
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "force_send_fields",
+				Description: "ForceSendFields is a list of field names (e.g. 'Combiner') to unconditionally include in API requests.",
+				Type:        proto.ColumnType_JSON,
+			},
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:74]: Not needed
```suggestion
```

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+			// standard steampipe columns
+			{
+				Name:        "tags",
+				Description: ColumnDescriptionTags,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("UserLabels"),
+			},
+			{
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:98]: Add `title` before `tags` column

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+				Description: ColumnDescriptionLocation,
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromConstant("global"),
+			},
+			{
+				Name:        "project",
+				Description: ColumnDescriptionProject,
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getProject,
+				Transform:   transform.FromValue(),
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:124]: Can we use `name` to get the project?

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	return req, nil
+}
+
+func getMonitoringAlertPolicyAka(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	alertPolicy := h.Item.(*monitoring.AlertPolicy)
+	
+	akas := []string{"gcp://monitoring.googleapis.com/" + alertPolicy.Name}
+
+	return akas, nil
+}
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:198]: Do we need hydrate function to build the aka? or, we can use transform function?

File: docs/tables/gcp_monitoring_alert_policy.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# Table:  gcp_monitoring_alert_policy
```
Comment (Subhajit97) [inline docs/tables/gcp_monitoring_alert_policy.md:1]:
```suggestion
# Table: gcp_monitoring_alert_policy
```

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@Paulami30 There are few changes required. Please see comments, thanks!
Can you also add the queries for the recommendations mentioned in the issue?

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+				Description: "The resource name for this policy.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "combiner",
+				Description: "How to combine the results of multiple conditions to determine if an incident should be opened.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "enabled",
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:40]: Add this column below `name`.

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+				Description: "Whether or not the policy is enabled.",
+				Type:        proto.ColumnType_BOOL,
+			},
+			{
+				Name:        "mutation_record",
+				Description: "A read-only record of the most recent change to the alerting policy.",
+				Type:        proto.ColumnType_STRING,
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:47]:
```suggestion
				Type:        proto.ColumnType_JSON,
```

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+				Name:        "tags",
+				Description: ColumnDescriptionTags,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("UserLabels"),
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromP(getMonitoringAlertPolicyAka, "Akas"),
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:97]: Do we need `FromP` to get a single property value?

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+	// Get project details
+	projectData, err := activeProject(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+	project := projectData.Project
+
+	resp := service.Projects.AlertPolicies.List("projects/" + project)
+
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:135]:
```suggestion
```

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	project := projectData.Project
+	name := d.KeyColumnQuals["name"].GetStringValue()
+
+	// Create Service Connection
+	service, err := MonitoringService(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := service.Projects.AlertPolicies.Get("projects/" + project + "/alertPolicies/" + name).Do()
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:167]: Is it correct?
As per the api `Name` includes both projectID and resourceID
image

File: gcp/table_gcp_monitoring_alert_policy.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+		return nil, nil
+	}
+
+	return req, nil
+}
+
+func getMonitoringAlertPolicyAka(_ context.Context, d *transform.TransformData) (interface{}, error) {
+	alertPolicy := d.HydrateItem.(*monitoring.AlertPolicy)
+
+	Akas := []string{"gcp://monitoring.googleapis.com/" + alertPolicy.Name}
```
Comment (Subhajit97) [inline gcp/table_gcp_monitoring_alert_policy.go:183]:
```suggestion
	akas := []string{"gcp://monitoring.googleapis.com/" + alertPolicy.Name}
```

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**